// stream-chat-angular/esm2020/lib/voice-recorder/audio-recorder.service.mjs
import { Injectable, NgModule } from '@angular/core';
import { MultimediaRecorder } from './media-recorder';
import { resampleWaveForm } from '../wave-form-sampler';
import * as i0 from "@angular/core";
import * as i1 from "../notification.service";
import * as i2 from "../chat-client.service";
import * as i3 from "./transcoder.service";
import * as i4 from "./amplitude-recorder.service";
/**
 * The `AudioRecorderService` can record an audio file, the SDK uses this to record a voice message
 */
export class AudioRecorderService extends MultimediaRecorder {
    constructor(notificationService, chatService, transcoder, amplitudeRecorder) {
        super(notificationService, chatService, transcoder);
        this.amplitudeRecorder = amplitudeRecorder;
        /**
         * Due to browser restrictions the following config is used:
         * - In Safari we record in audio/mp4
         * - For all other browsers we use audio/webm (which is then transcoded to wav)
         */
        this.config = {
            mimeType: MediaRecorder.isTypeSupported('audio/webm')
                ? 'audio/webm'
                : 'audio/mp4;codecs=mp4a.40.2', // fallback for Safari
        };
    }
    enrichWithExtraData() {
        const waveformData = resampleWaveForm(this.amplitudeRecorder.amplitudes, this.amplitudeRecorder.config.sampleCount);
        return { waveform_data: waveformData };
    }
    /**
     * Start audio recording
     */
    async start() {
        const result = await super.start();
        if (this.mediaRecorder?.stream) {
            this.amplitudeRecorder.start(this.mediaRecorder?.stream);
        }
        return result;
    }
    /**
     * Pause audio recording, it can be restarted using `resume`
     */
    pause() {
        const result = super.pause();
        this.amplitudeRecorder.pause();
        return result;
    }
    /**
     * Resume a previously paused recording
     */
    resume() {
        const result = super.resume();
        this.amplitudeRecorder.resume();
        return result;
    }
    /**
     * Stop the recording and free up used resources
     * @param options
     * @param options.cancel if this is `true` no recording will be created, but resources will be freed
     * @returns the recording
     */
    async stop(options) {
        try {
            const result = await super.stop(options);
            return result;
        }
        finally {
            this.amplitudeRecorder.stop();
        }
    }
}
AudioRecorderService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: AudioRecorderService, deps: [{ token: i1.NotificationService }, { token: i2.ChatClientService }, { token: i3.TranscoderService }, { token: i4.AmplitudeRecorderService }], target: i0.ɵɵFactoryTarget.Injectable });
AudioRecorderService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: AudioRecorderService, providedIn: NgModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: AudioRecorderService, decorators: [{
            type: Injectable,
            args: [{ providedIn: NgModule }]
        }], ctorParameters: function () { return [{ type: i1.NotificationService }, { type: i2.ChatClientService }, { type: i3.TranscoderService }, { type: i4.AmplitudeRecorderService }]; } });
