// stream-chat-angular/esm2020/lib/voice-recorder/voice-recorder.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VoiceRecorderComponent } from './voice-recorder.component';
import { VoiceRecordingModule } from '../voice-recording/voice-recording.module';
import { IconModule } from '../icon/icon.module';
import { TranslateModule } from '@ngx-translate/core';
import { AudioRecorderService } from './audio-recorder.service';
import { TranscoderService } from './transcoder.service';
import { AmplitudeRecorderService } from './amplitude-recorder.service';
import { VoiceRecorderWavebarComponent } from './voice-recorder-wavebar/voice-recorder-wavebar.component';
import * as i0 from "@angular/core";
export class VoiceRecorderModule {
}
VoiceRecorderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecorderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
VoiceRecorderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecorderModule, declarations: [VoiceRecorderComponent, VoiceRecorderWavebarComponent], imports: [CommonModule, VoiceRecordingModule, IconModule, TranslateModule], exports: [VoiceRecorderComponent, VoiceRecorderWavebarComponent] });
VoiceRecorderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecorderModule, providers: [
        AudioRecorderService,
        TranscoderService,
        AmplitudeRecorderService,
    ], imports: [CommonModule, VoiceRecordingModule, IconModule, TranslateModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecorderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [VoiceRecorderComponent, VoiceRecorderWavebarComponent],
                    imports: [CommonModule, VoiceRecordingModule, IconModule, TranslateModule],
                    exports: [VoiceRecorderComponent, VoiceRecorderWavebarComponent],
                    providers: [
                        AudioRecorderService,
                        TranscoderService,
                        AmplitudeRecorderService,
                    ],
                }]
        }] });
