// stream-chat-angular/esm2020/lib/message-input/voice-recorder.service.mjs
import { Injectable, NgModule } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * The `VoiceRecorderService` provides a commincation outlet between the message input and voice recorder components.
 */
export class VoiceRecorderService {
    constructor() {
        /**
         * Use this property to get/set if the recording component should be visible
         */
        this.isRecorderVisible$ = new BehaviorSubject(false);
        /**
         * The audio recording that was created
         */
        this.recording$ = new BehaviorSubject(undefined);
    }
}
VoiceRecorderService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecorderService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
VoiceRecorderService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecorderService, providedIn: NgModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecorderService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: NgModule,
                }]
        }], ctorParameters: function () { return []; } });
