// stream-chat-angular/esm2020/lib/message-reactions.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./chat-client.service";
import * as i2 from "./notification.service";
/**
 * The `MessageReactionsService` provides customization options to message [reactions](/chat/docs/javascript/send_reaction/).
 *
 */
export class MessageReactionsService {
    constructor(chatClientService, notificationService) {
        this.chatClientService = chatClientService;
        this.notificationService = notificationService;
        /**
         * The enabled [reactions](/chat/docs/javascript/send_reaction/) and the associated emoji
         *
         * You can provide any string as a reaction. The emoji can be provided as a string, if you want to use custom images for reactions you have to provide a [custom reactions UI](/chat/docs/sdk/angular/services/CustomTemplatesService/#messagereactionstemplate/)
         */
        this.reactions$ = new BehaviorSubject({
            haha: '😂',
            like: '👍',
            love: '❤️',
            sad: '😞',
            wow: '😮',
        });
    }
    /**
     * Sets the enabled reactions
     */
    set reactions(reactions) {
        this.reactions$.next(reactions);
    }
    /**
     * Get the currently enabled reactions
     */
    get reactions() {
        return this.reactions$.getValue();
    }
    /**
     * Query reactions of a specific message, more info in the [API documentation](/chat/docs/javascript/send_reaction/#query-reactions)
     * @param messageId
     * @param type
     * @param next
     * @returns the reactions and the cursor for the next/prev pages
     */
    async queryReactions(messageId, type, next) {
        if (!this.chatClientService.chatClient) {
            throw new Error('Intialize the ChatClientService before querying reactions');
        }
        else {
            try {
                const response = await this.chatClientService.chatClient.queryReactions(messageId, { type }, { created_at: -1 }, { next });
                return response;
            }
            catch (error) {
                this.notificationService.addTemporaryNotification('streamChat.Error loading reactions');
                throw error;
            }
        }
    }
}
MessageReactionsService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: MessageReactionsService, deps: [{ token: i1.ChatClientService }, { token: i2.NotificationService }], target: i0.ɵɵFactoryTarget.Injectable });
MessageReactionsService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: MessageReactionsService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: MessageReactionsService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.ChatClientService }, { type: i2.NotificationService }]; } });
