// stream-chat-angular/esm2020/lib/message.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * The message service contains configuration options related to displaying the message content
 */
export class MessageService {
    constructor() {
        /**
         * Decides if the message content should be formatted as text or HTML
         *
         * If you display messages as text the following parts are still be displayed as HTML:
         * - user mentions -> you can customize this with your own template using the [`customTemplatesService.mentionTemplate$`](/chat/docs/sdk/angular/services/CustomTemplatesService/#mentiontemplate)
         * - links -> you can customize this by providing you own [`customLinkRenderer`](#customlinkrenderer) method
         */
        this.displayAs = 'text';
    }
    /**
     * Tells if an attachment is custom (you need to provide your own template to display them) or built-in (the SDK supports it out-of-the-box)
     * @param attachment
     * @returns `true` if the attachment is custom
     */
    isCustomAttachment(attachment) {
        if (this.filterCustomAttachment) {
            return this.filterCustomAttachment(attachment);
        }
        else {
            return (attachment.type !== 'image' &&
                attachment.type !== 'file' &&
                attachment.type !== 'video' &&
                attachment.type !== 'voiceRecording' &&
                attachment.type !== 'giphy');
        }
    }
}
MessageService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: MessageService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
MessageService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: MessageService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: MessageService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return []; } });
