// stream-chat-angular/esm2020/lib/voice-recording/voice-recording.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VoiceRecordingComponent } from './voice-recording.component';
import { VoiceRecordingWavebarComponent } from './voice-recording-wavebar/voice-recording-wavebar.component';
import { IconModule } from '../icon/icon.module';
import { TranslateModule } from '@ngx-translate/core';
import * as i0 from "@angular/core";
export class VoiceRecordingModule {
}
VoiceRecordingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecordingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
VoiceRecordingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecordingModule, declarations: [VoiceRecordingComponent, VoiceRecordingWavebarComponent], imports: [CommonModule, IconModule, TranslateModule], exports: [VoiceRecordingComponent, VoiceRecordingWavebarComponent] });
VoiceRecordingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecordingModule, imports: [CommonModule, IconModule, TranslateModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: VoiceRecordingModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [VoiceRecordingComponent, VoiceRecordingWavebarComponent],
                    imports: [CommonModule, IconModule, TranslateModule],
                    exports: [VoiceRecordingComponent, VoiceRecordingWavebarComponent],
                }]
        }] });
