// angular-mentions/esm2020/lib/mention.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MentionDirective } from './mention.directive';
import { MentionListComponent } from './mention-list.component';
import * as i0 from "@angular/core";
export class MentionModule {
}
MentionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: MentionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MentionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: MentionModule, declarations: [MentionDirective,
        MentionListComponent], imports: [CommonModule], exports: [MentionDirective] });
MentionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: MentionModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: MentionModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MentionDirective,
                        MentionListComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        MentionDirective
                    ]
                }]
        }] });
